File: backend/routes/financial_math.py
from flask import Blueprint, request, jsonify

financial_math_module = Blueprint('financial_math_module', __name__)

# Future Value Calculation
def future_value(principal, rate, time):
    return principal * (1 + rate) ** time

# Present Value Calculation
def present_value(future_value, rate, time):
    return future_value / (1 + rate) ** time

# Compound Interest Calculation
def compound_interest(principal, rate, time):
    return principal * ((1 + rate) ** time - 1)

# Annuity Calculation
def annuity(payment, rate, time):
    return payment * (((1 + rate) ** time - 1) / rate)

# Route to handle financial math calculations
@financial_math_module.route('/calculate', methods=['POST'])
def calculate_financial_math():
    data = request.json
    principal = float(data.get('principal', 0))
    rate = float(data.get('rate', 0)) / 100  # Convert percentage to decimal
    time = float(data.get('time', 0))
    calculation_type = data.get('calculation_type')

    result = None

    if calculation_type == 'future_value':
        result = future_value(principal, rate, time)
    elif calculation_type == 'present_value':
        # Assume future value is equal to principal for simplicity in this example
        result = present_value(principal, rate, time)
    elif calculation_type == 'compound_interest':
        result = compound_interest(principal, rate, time)
    elif calculation_type == 'annuity':
        # Assume payment is equal to principal for simplicity in this example
        result = annuity(principal, rate, time)
    else:
        return jsonify({"error": "Invalid calculation type"}), 400

    return jsonify({"result": result}), 200

File: frontend/jinja2_frontend/static/financial_math.js
document.getElementById('financial-math-form').addEventListener('submit', async function(event) {
    event.preventDefault();

    const formData = {
        principal: document.getElementById('principal').value,
        rate: document.getElementById('rate').value,
        time: document.getElementById('time').value,
        calculation_type: document.getElementById('calculation-type').value
    };

    // Add payment for annuity calculations
    if (formData.calculation_type === 'future_value_annuity' || formData.calculation_type === 'present_value_annuity') {
        formData.payment = document.getElementById('payment').value;
    }

    // Add compounding frequency for compound interest
    if (formData.calculation_type === 'compound_interest') {
        formData.compounding_frequency = document.getElementById('compounding_frequency').value;
    }

    try {
        const response = await fetch('/financial_math/calculate', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(formData)
        });

        if (!response.ok) {
            throw new Error('Failed to calculate financial math.');
        }

        const result = await response.json();
        document.getElementById('result').innerHTML = `<h2>Result: $${result.result.toFixed(2)}</h2>`;
        document.getElementById('result').style.display = "block";
    } catch (error) {
        console.error('Error calculating financial math:', error);
        document.getElementById('result').innerHTML = `<p>Error: ${error.message}</p>`;
    }
});

// Show/hide input fields based on calculation type
document.getElementById('calculation-type').addEventListener('change', function() {
    const calcType = this.value;

    // Show payment field for annuity calculations
    if (calcType === 'future_value_annuity' || calcType === 'present_value_annuity') {
        document.getElementById('payment-field').style.display = 'block';
    } else {
        document.getElementById('payment-field').style.display = 'none';
    }

    // Show compounding frequency for compound interest
    if (calcType === 'compound_interest') {
        document.getElementById('compounding-field').style.display = 'block';
    } else {
        document.getElementById('compounding-field').style.display = 'none';
    }
});
